apps/meteor/app/livechat/server/api/v1/visitor.ts

```diff
@@ -68,16 +68,41 @@ API.v1.addRoute('livechat/visitor', {
 			);
 		}
 
-		if (customFields && Array.isArray(customFields)) {
-			for await (const field of customFields) {
-				const customField = await LivechatCustomField.findOneById(field.key);
-				if (!customField) {
-					continue;
-				}
-				const { key, value, overwrite } = field;
-				if (customField.scope === 'visitor' && !(await VisitorsRaw.updateLivechatDataByToken(token, key, value, overwrite))) {
-					return API.v1.failure();
-				}
+		if (customFields && Array.isArray(customFields) && customFields.length > 0) {
+			const keys = customFields.map((field) => field.key);
+			const errors: string[] = [];
+
+			const processedKeys = await Promise.all(
+				await LivechatCustomField.findByIdsAndScope<Pick<ILivechatCustomField, '_id'>>(keys, 'visitor', {
+					projection: { _id: 1 },
+				})
+					.map(async (field) => {
+						const customField = customFields.find((f) => f.key === field._id);
+						if (!customField) {
+							return;
+						}
```

Comment:
Do we really need this? 🤔 

Is there any case that the values from the db will not match the values provided by the user? I'm asking because we are fetching that by custom field keys, in theory, this will never fail.

To me it would make more sense to check the keys.length against the length of the array returned from the DB, if the keys are greater than the values returned from the DB, then yes, we could assume we have some key in the array, that is not present on the DB.

Maybe I'm missing something.

Reply:
This is for getting the value that the user sent. We fetch the CF from the DB and that will give us the value that's already stored on the DB. Then we find that on the array of values from the user and use that value.

`find` returns T | null so the if is necessary.